### PR TITLE
[F] Typography Rules and Mixins

### PR DESCRIPTION
**Pull request by drapisarda**
https://homeday.atlassian.net/browse/DS-97

Implemented a mixin retrieve the typography styles documented here 
https://www.figma.com/file/KdznSjw9btqHEKdgmm2UT4co/Design-System-UI-Patterns?node-id=572%3A1454

Also, there's a set of CSS rules for a full typography definition, styling paragraphs and headings.
The styles are taken from the DS font style sets. The spacing between elements is relative to the font size of the element itself. This can be changed based on the outcome of the next designers' sync meeting.

WIP:
The margin between paragraphs and headings are yet to be confirmed.

### CODE REVIEW

**Review by volcanioo — APPROVED** [review]

File: src/stories/Typography.stories.js
Comment (kuroski) [inline src/stories/Typography.stories.js:7]: Please, don't forget to use the new Storybook syntax =D

You can find more information on this PR:
- https://github.com/homeday-de/homeday-blocks/pull/512

Or by looking at some components that already are using the new format:
- https://github.com/homeday-de/homeday-blocks/blob/develop/src/stories/HdButton.stories.js
- https://github.com/homeday-de/homeday-blocks/blob/develop/src/stories/HdDashedList.stories.js
- https://github.com/homeday-de/homeday-blocks/blob/develop/src/stories/HdNotification.stories.js

If you need any help, just ping me 😄

**Review by kuroski — COMMENTED** [review]
Great work 😄😄😄

Just one comment regarding the documentation =D

File: src/styles/typography.scss
Comment (ilyasmez) [inline src/styles/typography.scss:97]: This whole logic is a little bit disconnected from our design system because it doesn't make use of the `font("DS-N")` mixin, it's there just to allow using other scales.
I suggest to remove the `document-typography` mixing but keep the style definitions of the headings and the paragraph element, that can be useful to have some styles our of the box, bu they should use the `font("DS-N")` mixin, e.g.
```scss
h1 {
  @include font("DS-800");
}
```

And I don't think we need to touch the margins.

**Review by ilyasmez — CHANGES_REQUESTED** [review]
Nice work Daniele 👏 

Please check my comments, and let's have a conversation there.

---

I also think it will be useful if you add `font-family: $font-primary;` in [main.scss#L12](https://github.com/homeday-de/homeday-blocks/blob/develop/src/styles/main.scss#L12), to set the default font family ("Source Sans Pro") of our apps and storybook, that will make your demo similar to the design.

---

For the rest of the team, please have a thoughtful look at the PR, and see if it makes sense to you and matches your expectations, because this is something that we will probably be using for a couple of years 🤞

File: src/stories/styles/Typography.scss
Comment (drapisarda) [reply src/stories/styles/Typography.scss:29]: Yes. That's why I migrated those in the stories files, because those are meant to be for demo and documentation purpose only.

File: src/stories/styles/Typography.scss
Comment (drapisarda) [reply src/stories/styles/Typography.scss:69]: You're right. I'm going to put this in a separate scss file, so the user can import the mixins without the need of importing all those classes

File: src/styles/typography.scss
Comment (drapisarda) [reply src/styles/typography.scss:97]: Don't you think it's a nice to have for a fast prototyping?

File: src/stories/Typography.stories.js
Comment (drapisarda) [reply src/stories/Typography.stories.js]: Done!

**Review by kuroski — APPROVED** [review]
Great work 😄😄 
After @ilyasmez request are done :shipit:

**Review by leandroinacio — APPROVED** [review]

**Review by volcanioo — APPROVED** [review]

**Comment by ilyasmez** [comment]
Thank you @drapisarda for taking care of the adjustments, looks better 👍 

Regarding the `document-typography` I still think it doesn't belong to our design system, it's completely not related to it (doesn't use `DS-N` or redefine it, it just sets the `font-size` directly for the heading and text elements) and that will just cause conflicts, and bring unnecessary complexity to the implementation that we will have to maintain later.
You mentioned that it can be useful for prototyping, well, for that you can still use our scale instead of the predefined ones, as we want to keep the same scale across all of our products. What I meant in [that comment](https://github.com/homeday-de/homeday-blocks/pull/587#discussion_r506296045) is that you can actually use the `DS-N` presets we have to style the heading and text elements out of the box, especially because we are using `reset.css` so it's better to provide some styles already when semantic html is used.

I think what we need is just:
- [ x ] The `font("DS-N")` mixin definition
- [ x ] A set of class names that make use of the `font("DS-N")` mixin in case the developer prefers to use classnames.
- [ ] A story to demonstrate the different fonts we have (can be a 'lorem ipsum', but we have to mention which preset is used for each element, e.g. "DS-100: Lorem ipsum dolor sit amet.").
- [ ] Assign the new styles to the native text and heading elements to provide the projects with some styles out of the box.
(e.g. `h1 { @include font("DS-800") } h2 { @include font("DS-700") }...`, and we don't need the `.text_small` btw)

If I'm still not clear enough please let me know and we can have a call to clarify things.

**Review by ilyasmez — COMMENTED** [review]
Looks much better and cleaner 👍 

Here are some points that I think we have to take care of before merging:

----- 
#### Deprecation of the old `font`
Now we have two `font` mixins, [the old](https://github.com/homeday-de/homeday-blocks/blob/4ef782f1311f1c98b9776685f8609b36e081a4e8/src/styles/_fonts.scss#L1) and [the new one](https://github.com/homeday-de/homeday-blocks/blob/4ef782f1311f1c98b9776685f8609b36e081a4e8/src/styles/typography.scss#L8).
Your changes will break the old `font` mixin, and it will force all the squads to migrate to the new font mixin when upgrading (doesn't seem like a good DX).

I suggest renaming the old `font` mixin to `deprecated_font`, and use it as a fall back in the new `font`, like:

```scss
@import './deprecated_font';

@mixin font($style) {
  @if ($style == "DS-60") {
    ...
  }
  @else if ($style == "DS-80") {
    ...
  }
  else deprecated_font($style);
}
```

And in `deprecate_font` we add a deprecation warning:
```scss
@mixin deprecated_font($style) {
  @warn "Please use the new 'font' mixin, link_to_the_mapping_guide";

  @if ($style == "headline") {
    ...
  }
  ...
```

-----
#### Font family

We have to stop using the old font in our code, for example in `main.scss` we should not use `font('body')`, we can just remove it and explicitly add `font-family: $primary-font;` (that will fix the font in Storybook as well).

-----
#### Components' dependency

There are some components that depend on the old mixin, like `HdLink` (that's the only one I can think of), for that we will have to move it's styles inside `HdLink.vue`

-----

If you think that the scope is getting bigger for one PR (I would 100% agree), we can also merge this PR in a separate branch that we can use to improve the implementation, and once it's done we merge everything to `develop`. Let me know what you think.

File: src/stories/TypographyHomeday.stories.js
Comment (drapisarda) [reply src/stories/TypographyHomeday.stories.js:24]: I set the outline to enlight better the margins from the typography implemented, avoiding to confuse it with the storybook styling.

File: src/styles/mixins.scss
Comment (drapisarda) [reply src/styles/mixins.scss:20]: Nice

**Comment by drapisarda** [comment]
@ilyasmez what about :
main.scss 
`body {
     @include hd-typography();
     color: $primary-color;
}`

typography.scss
`@mixin hd-typography() {
     font-family: $font-primary;
....`
?

File: src/stories/TypographyHomeday.stories.js
Comment (ilyasmez) [reply src/stories/TypographyHomeday.stories.js]: Ah I see, it just looked like some outline you used for debugging. It looks great now with the background color instead 👍

**Comment by ilyasmez** [comment]
@drapisarda That's some awesome job man 👏 👏 

What do you think of my suggestion regarding `HdLink`? Do you think you can move the styles inside the component in this PR?
And what about the `margin-bottom` and `margin-top` you added to the heading and paragraph elements, are they specified by the design team somewhere?

**Comment by drapisarda** [comment]
> @drapisarda That's some awesome job man 👏 👏
> 
> What do you think of my suggestion regarding `HdLink`? Do you think you can move the styles inside the component in this PR?
> And what about the `margin-bottom` and `margin-top` you added to the heading and paragraph elements, are they specified by the design team somewhere?

For HDLink I would keep it in another PR and address it ASAP. 
About the margins, I'm waiting for feedback from the designers on the right margins to use.

**Comment by ilyasmez** [comment]
> For HDLink I would keep it in another PR and address it ASAP. Also, I that element is independent from the typography as the link itself inherits line-height and font-size from its context.

That makes sense. Please create a ticket in the DS board so we don't forget about it.

> About the margins, I'm waiting for feedback from the designers on the right margins to use.

Ah, I see, then I'd suggest to just remove the margins for now, and when we align with the design team we add the right ones, because the margins are also a topic on their own.

**Review by ilyasmez — APPROVED** [review]
Let's fly! 🚀
Great job! 👏 

Just check my last comment, and we're good to merge.

**Comment by drapisarda** [comment]
I commented out the margins to force the idea that those have to be defined.

**Pull request merged**
